pinot-server/src/main/java/org/apache/pinot/server/predownload/PredownloadScheduler.java

```diff
@@ -0,0 +1,430 @@
+  private static final long LOAD_SEGMENTS_TIMEOUT_MIN = 5;
+  private final PropertiesConfiguration _properties;
+  private final PinotConfiguration _pinotConfig;
+  private final InstanceDataManagerConfig _instanceDataManagerConfig;
+  private final String _clusterName;
+  private final String _instanceId;
+  private final String _zkAddress;
+  @VisibleForTesting
+  Executor _executor;
+  @VisibleForTesting
+  Set<String> _failedSegments;
+  private PredownloadMetrics _predownloadMetrics;
+  private int _numOfSkippedSegments;
+  private int _numOfUnableToDownloadSegments;
+  private int _numOfDownloadSegments;
+  private long _totalDownloadedSizeBytes;
+  private PredownloadZKClient _predownloadZkClient;
+  private List<PredownloadSegmentInfo> _predownloadSegmentInfoList;
+  private Map<String, PredownloadTableInfo> _tableInfoMap;
+
+  public PredownloadScheduler(PropertiesConfiguration properties)
+      throws Exception {
+    _properties = properties;
+    _clusterName = properties.getString(CommonConstants.Helix.CONFIG_OF_CLUSTER_NAME);
+    _zkAddress = properties.getString(CommonConstants.Helix.CONFIG_OF_ZOOKEEPR_SERVER);
+    _instanceId = properties.getString(CommonConstants.Server.CONFIG_OF_INSTANCE_ID);
+    _pinotConfig = new PinotConfiguration(properties);
+    _instanceDataManagerConfig =
+        new HelixInstanceDataManagerConfig(new ServerConf(_pinotConfig).getInstanceDataManagerConfig());
+    // Get the number of available processors (vCPUs)
+    int numProcessors = Runtime.getRuntime().availableProcessors();
+    _failedSegments = ConcurrentHashMap.newKeySet();
+    // TODO: tune the value
+    _executor = Executors.newFixedThreadPool(numProcessors * 3);
+    LOGGER.info("Created thread pool with num of threads: {}", numProcessors * 3);
+    _numOfSkippedSegments = 0;
+    _numOfDownloadSegments = 0;
+  }
+
+  public void start() {
+
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        try {
+          LOGGER.info("Trying to stop predownload process!");
+          stop();
+        } catch (Exception e) {
+          LOGGER.error("error shutting down predownload process : ", e);
+        }
+      }
+    });
+
+    long startTime = System.currentTimeMillis();
+    initializeZK();
+    initializeMetricsReporter();
+    initializeSegmentFetcher();
+    getSegmentsInfo();
+    loadSegmentsFromLocal();
+    PredownloadCompletionReason reason = downloadSegments();
+    long timeTaken = System.currentTimeMillis() - startTime;
+    LOGGER.info(
+        "Predownload process took {} sec, tried to download {} segments, skipped {} segments "
+            + "and unable to download {} segments. Download size: {} MB. Download speed: {} MB/s",
+        timeTaken / 1000, _numOfDownloadSegments, _numOfSkippedSegments, _numOfUnableToDownloadSegments,
+        _totalDownloadedSizeBytes / (1024 * 1024),
+        (_totalDownloadedSizeBytes / (1024 * 1024)) / (timeTaken / 1000 + 1));
```

Comment:
If _totalDownloadedSizeBytes is small, this can result in 0 MB/s. shall we use float points?